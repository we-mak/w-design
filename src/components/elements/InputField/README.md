| Props        | type                                             | default    | description                   |
| ------------ | ------------------------------------------------ | ---------- | ----------------------------- |
| inputRef     | React.MutableRefObject                           |            | input refs                    |
| appearance   | string                                           | "standard" |                               |
| type         | string                                           | "text"     | type of input                 |
| name         | string                                           |            | field name                    |
| className    | string                                           |            | class name of input field     |
| autoComplete | string                                           | "off"      | control browser auto complete |
| size         | string                                           |            |                               |
| tabIndex     | number                                           |            | tabindex order                |
| value        | string                                           |            |                               |
| defaultValue | string                                           |            |                               |
| isRequired   | boolean                                          |            |                               |
| isError      | boolean                                          |            |                               |
| isValid      | boolean                                          |            |                               |
| isWarning    | boolean                                          |            |                               |
| isFocus      | boolean                                          | false      |                               |
| isDisabled   | boolean                                          | false      |                               |
| onChange     | (e: React.ChangeEvent<HTMLInputElement>) => void |            |                               |
| onBlur       | (e: React.SyntheticEvent<HTMLElement>) => void   |            |                               |
