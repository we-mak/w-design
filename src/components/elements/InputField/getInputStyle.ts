import { css } from "styled-components";
import { InputFieldProps } from "./types";
import {
  getColor,
  hex2Rgba,
  getFontSize,
  getElementSize,
  padding,
  height
} from "../../../common/styleUtils/utils";

export const getInputStyle = (props: InputFieldProps) => {
  const {
    size = "md",
    fluid,
    isError,
    isValid,
    isWarning,
    theme: { colors, fontSizes, radii, lineHeights }
  } = props;

  const boderColor = () => {
    if (isWarning) return colors["O30"];
    if (isValid) return colors["G30"];
    if (isError) return colors["R30"];
    return colors["N5"];
  };

  const boderFocusColor = () => {
    if (isWarning) return colors["O30"];
    if (isValid) return colors["G30"];
    if (isError) return colors["R30"];
    return colors["B50"];
  };

  let width;
  if (fluid) width = "100%";
  const paddingStyle = getElementSize(padding, size);
  const fontSizeStyle = getFontSize(size, fontSizes);
  const heightStyle = getElementSize(height, size);

  return css`
    background: transparent;
    background-image: none;
    outline: none;
    max-width: 100%;
    display: block;
    position: relative;
    border: 0.05rem solid ${boderColor()};
    border-radius: ${radii[2]};
    color: ${colors["N70"]};
    padding: ${paddingStyle};
    font-size: ${fontSizeStyle};
    height: ${heightStyle};
    width: ${width};
    line-height: ${lineHeights[1]};
    transition: background 0.2s, border 0.2s, box-shadow 0.2s, color 0.2s;

    &:focus {
      box-shadow: 0 0 0 0.1rem ${hex2Rgba(boderFocusColor(), 0.2)};
      border-color: ${boderFocusColor()};
    }

    &::placeholder {
      color: ${colors["N7"]};
    }

    &:disabled,
    &[disabled] {
      cursor: not-allowed;
      color: ${getColor("N7", colors)} !important;
      background: ${getColor("N2", colors)} !important;
    }

    &[type="file"] {
      height: auto;
    }
  `;
};
