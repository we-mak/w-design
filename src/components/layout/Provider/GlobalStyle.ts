import { createGlobalStyle } from "styled-components";
import theme from "../../../common/styleUtils/theme";
const colors = theme.colors;

/*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css */
export const GlobalStyle = createGlobalStyle`
html {
  line-height: 1.15;
  -webkit-text-size-adjust: 100%;
}
body {
  margin: 0;
}
main {
  display: block;
}
h1 {
  font-size: 2em;
  margin: 0.67em 0;
}
hr {
  box-sizing: content-box;
  height: 0;
  overflow: visible;
}
pre {
  font-family: monospace, monospace;
  font-size: 1em;
}
a {
  background-color: transparent;
}
abbr[title] {
  border-bottom: none;
  text-decoration: underline;
  text-decoration: underline dotted;
}
b,
strong {
  font-weight: bolder;
}
code,
kbd,
samp {
  font-family: monospace, monospace;
  font-size: 1em;
}
small {
  font-size: 80%;
}
sub,
sup {
  font-size: 75%;
  line-height: 0;
  position: relative;
  vertical-align: baseline;
}
sub {
  bottom: -0.25em;
}
sup {
  top: -0.5em;
}
img {
  border-style: none;
}
button,
input,
optgroup,
select,
textarea {
  font-family: inherit;
  font-size: 100%;
  line-height: 1.15;
  margin: 0;
}
button,
input {
  overflow: visible;
}
button,
select {
  text-transform: none;
}
[type="button"],
[type="reset"],
[type="submit"],
button {
  -webkit-appearance: button;
}
[type="button"]::-moz-focus-inner,
[type="reset"]::-moz-focus-inner,
[type="submit"]::-moz-focus-inner,
button::-moz-focus-inner {
  border-style: none;
  padding: 0;
}
[type="button"]:-moz-focusring,
[type="reset"]:-moz-focusring,
[type="submit"]:-moz-focusring,
button:-moz-focusring {
  outline: 1px dotted ButtonText;
}
fieldset {
  padding: 0.35em 0.75em 0.625em;
}
legend {
  box-sizing: border-box;
  color: inherit;
  display: table;
  max-width: 100%;
  padding: 0;
  white-space: normal;
}
progress {
  vertical-align: baseline;
}
textarea {
  overflow: auto;
}
[type="checkbox"],
[type="radio"] {
  box-sizing: border-box;
  padding: 0;
}
[type="number"]::-webkit-inner-spin-button,
[type="number"]::-webkit-outer-spin-button {
  height: auto;
}
[type="search"] {
  -webkit-appearance: textfield;
  outline-offset: -2px;
}
[type="search"]::-webkit-search-decoration {
  -webkit-appearance: none;
}
::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}
details {
  display: block;
}
summary {
  display: list-item;
}
template {
  display: none;
}
[hidden] {
  display: none;
}

*,
*::before,
*::after {
  box-sizing: inherit;
}

html {
  box-sizing: border-box;
  font-size: 20px;
  line-height: 1.5;
  -webkit-tap-highlight-color: transparent;
}

body {
  font-size: .8rem;
  overflow-x: hidden;
  text-rendering: optimizeLegibility;
}


h1 {
    font-size: 1.6rem;
  }

  h2 {
    font-size: 1.4rem;
  }

  h3 {
    font-size: 1.2rem;
  }

  h4 {
    font-size: 1rem;
  }
  h5 {
    font-size: 0.9rem;
  }

  h6 {
    font-size: 0.8rem;
  }

  h1,
  h2,
  h3,
  h4,
  h5,
  h6 {
    font-weight: 700;
    line-height: 1.25;
    margin-bottom: 0.5em;
    margin-top: 0;
    font-family: ${(props: any) => props.theme.fonts.medium};
  }

  h1,
  h2,
  h3,
  h4 {
    color: ${colors.N50};
  }

  h5,
  h6 {
    color: ${colors.N60};
  }

  p {
    color: ${colors.N70};
    line-height: 1.5;
    margin-bottom: 0.5em;
    margin-top: 0;
    font-family: ${(props: any) => props.theme.fonts.normal};
  }

  a {
    font-family: ${(props: any) => props.theme.fonts.normal};
    color: ${colors.B50};
    outline: none;
    text-decoration: none;

    &:hover {
      color: ${colors.B30};
    }

    &:focus,
    &:hover,
    &:active {
      text-decoration: underline;
    }
    &:focus,
    &:active {
      color: ${colors.B60};
    }
    &:focus {
      text-decoration: none;
    }
  }

  b,
  strong {
    font-weight: bolder;
  }

  small {
    font-size: 80%;
    font-weight: 400;
    color: ${colors.N30};
  }

  blockquote {
    margin-left: 0;
    padding: 0.4rem 0.8rem;
    color: ${colors.N30};
    p:last-child {
      margin-bottom: 0;
    }
  }

  table {
    border-collapse: collapse;
    border-spacing: 0;
    text-align: left;
    width: 100%;
  }
  th {
    border-bottom: 0.05rem solid ${props => props.theme.colors["N5"]};
    padding: 0.6rem 0.4rem;
    border-bottom-width: 0.1rem;
  }
  td {
    border-bottom: 0.05rem solid ${props => props.theme.colors["N2"]};
    padding: 0.6rem 0.4rem;
  }
`;
