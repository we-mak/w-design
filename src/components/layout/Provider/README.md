Theme provider is based on [styled-components](https://www.styled-components.com/) and making theme with [styled-system](http://jxnblk.com/styled-system/)

Provider provides theme for your layout, you can supply your theme by insert your theme object by `theme` props
Default theme

```js static
defaultTheme = {
  breakpoint: ["24em", "32em", "48em", "64em", "80em"],
  fonts:
    '-apple-system, BlinkMacSystemFont, Helvetica, "Helvetica Neue", system-ui, sans-serif',
  fontSizes: [
    "0.7rem",
    "0.8rem",
    "0.9rem",
    "1rem",
    "1.2rem",
    "1.4rem",
    "1.6rem",
    "1.8rem",
    "2rem"
  ],
  fontWeights = { normal: 400, strong: 500, bold: 700, xbold: 900 },
  lineHeights = [1, 1.125, 1.25, 1.5],
  letterSpacings = { normal: "normal", caps: "0.25em" },
  radii = ["0rem", "0.1rem", "0.2rem", "0.3rem", "0.4rem"],
  space = [0, 4, 8, 16, 32, 64, 128], // space is used for margin and padding scales
  shadows = [
    "none",
    `0 .2rem .5rem ${hex2Rgba(colors.N50, 0.5)}`,
    `inset 0 0 0 1px rgba(87, 85, 217, .2), 0 0 4px rgba(87, 85, 217, .2)`
  ],
  colors: {...} // see colors
};
```

```js
<Provider>
  <h1>Default </h1>
  <p>
    Lorem ipsum
    <span>
      <a href="#">dolor </a>
    </span>
    sip amet
  </p>
</Provider>
```

```js
<Provider
  theme={{
    fonts: "Roboto, Helvetica, sans-serif"
  }}
>
  <h1>Roboto</h1>
  <p>
    Lorem ipsum
    <span>
      <a href="#">dolor </a>
    </span>
    sip amet
  </p>
</Provider>
```
