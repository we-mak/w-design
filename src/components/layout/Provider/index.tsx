import * as React from "react";
import { ThemeProvider } from "styled-components";
import defaultTheme from "../../../common/theme";
import { ResetCss } from "./ResetCss";
import Root from "./Root";

export interface Props extends React.HTMLProps<HTMLDivElement> {
  theme?: any;
}

class Provider extends React.Component<Props> {
  stylesheet: any;

  static displayName = "WProvider";

  public static defaultProps: Props = {
    theme: {}
  };

  componentDidMount() {
    this.stylesheet = document.createElement("style");
    this.stylesheet.type = "text/css";
    // this.stylesheet.setAttribute("rel", "preload");
    // this.stylesheet.setAttribute("as", "style");
    this.stylesheet.innerHTML = ResetCss;

    if (document && document.head) {
      document.head.appendChild(this.stylesheet);
    }
  }

  componentWillUnmount() {
    if (this.stylesheet && document && document.head) {
      document.head.removeChild(this.stylesheet);
      delete this.stylesheet;
    }
  }

  render() {
    const { theme, ...props } = this.props;
    return (
      <ThemeProvider theme={{ ...defaultTheme, ...theme }}>
        <Root {...props as any} />
      </ThemeProvider>
    );
  }
}

export default Provider;
