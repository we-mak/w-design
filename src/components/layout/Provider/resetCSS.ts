import { colors } from "../../../common/colors";

const ResetCss = `
*,
*::before,
*::after {
  box-sizing: inherit;
}

html {
  box-sizing: border-box;
  font-size: 20px;
  line-height: 1.5;
  -webkit-tap-highlight-color: transparent;
}

body {
  background: ${colors.WHITE};
  color: ${colors.N60};
  font-family: -apple-system, BlinkMacSystemFont, Helvetica, "Helvetica Neue", system-ui, sans-serif;
  font-size: 0.8rem;
  overflow-x: hidden;
  text-rendering: optimizeLegibility;
}

a {
  font-family: Helvetica, "Helvetica Neue", system-ui, sans-serif;
}
`;

export default ResetCss;
