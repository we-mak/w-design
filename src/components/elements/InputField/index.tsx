/**
 * Input Field
 * Single input field
 */
import * as React from "react";
import styled from "styled-components";
import { InputFieldProps } from "./types";
import { getInputStyle } from "./getInputStyle";

const StyledInput = styled.input<any>`
  ${getInputStyle};
`;
StyledInput.displayName = "StyledInput";

const InputField = (props: InputFieldProps) => {
  const {
    inputRef,
    appearance = "standard",
    type = "text",
    name,
    className,
    id,
    autoComplete = "off",
    size = "md",
    tabIndex,
    value,
    defaultValue,
    isRequired,
    isError,
    isValid,
    isWarning,
    isFocus = false,
    isDisabled = false,
    onChange,
    onBlur,
    ...rest
  } = props;

  let input!: HTMLInputElement | undefined;

  const getInputRef = (ref: HTMLInputElement) => {
    input = ref;
    if (inputRef) inputRef(ref);
  };

  React.useEffect(() => {
    if (isFocus && input) input.focus();
    return () => (input = undefined);
  });

  return (
    <>
      <StyledInput
        ref={getInputRef}
        type={type}
        size={size}
        name={name}
        appearance={appearance}
        autoComplete={autoComplete}
        className={className}
        id={id}
        required={isRequired}
        value={value}
        defaultValue={defaultValue}
        isFocus={isFocus}
        isValid={isValid}
        isError={isError}
        isWarning={isWarning}
        disabled={isDisabled}
        onChange={isDisabled ? undefined : onChange}
        onBlur={onBlur}
        tabIndex={tabIndex}
        {...rest}
      />
    </>
  );
};

export default InputField;
