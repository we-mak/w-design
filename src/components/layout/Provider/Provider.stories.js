import React from "react";
import { storiesOf } from "@storybook/react";
import Provider from "./Provider";

storiesOf("components/Provider", module).addWithJSX("default Provider", () => (
  <Provider>
    <h1>Default </h1>
    <p>
      Lorem ipsum
      <span>
        <a href="#"> dolor </a>
      </span>{" "}
      sip amet
    </p>
  </Provider>
));
