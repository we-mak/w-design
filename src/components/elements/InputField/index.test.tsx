import * as React from "react";
import { render, cleanup } from "@testing-library/react";
import "jest-styled-components";
import InputField from "./index";
import Provider from "../../layout/Provider";

afterEach(cleanup);

describe("<AvatarGroup/>", () => {
  // Render test
  it("should renders correctly", () => {
    const { container } = render(
      <Provider>
        <InputField placeholder="Default input" />
        <InputField size="lg" placeholder="Large input" />
        <InputField size="sm" placeholder="Small input" />
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });
});
