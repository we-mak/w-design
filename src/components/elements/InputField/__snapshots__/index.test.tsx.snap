// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AvatarGroup/> should renders correctly 1`] = `
.c1 {
  background: transparent;
  background-image: none;
  outline: none;
  max-width: 100%;
  display: block;
  position: relative;
  border: 0.05rem solid #BECEDC;
  border-radius: 0.2rem;
  color: #202B33;
  padding: 0.35rem 0.4rem;
  font-size: 0.8rem;
  font-family: inherit;
  height: 1.8rem;
  line-height: 1.125;
  -webkit-transition: background 0.2s,border 0.2s,box-shadow 0.2s,color 0.2s;
  transition: background 0.2s,border 0.2s,box-shadow 0.2s,color 0.2s;
}

.c1:focus {
  box-shadow: 0 0 0 0.1rem rgba(0,140,192,0.2);
  border-color: #008CC0;
}

.c1::-webkit-input-placeholder {
  color: #90A9BF;
}

.c1::-moz-placeholder {
  color: #90A9BF;
}

.c1:-ms-input-placeholder {
  color: #90A9BF;
}

.c1::placeholder {
  color: #90A9BF;
}

.c1:disabled,
.c1[disabled] {
  cursor: not-allowed;
  color: #90A9BF !important;
  background: #EFF3F6 !important;
}

.c1[type="file"] {
  height: auto;
}

.c2 {
  background: transparent;
  background-image: none;
  outline: none;
  max-width: 100%;
  display: block;
  position: relative;
  border: 0.05rem solid #BECEDC;
  border-radius: 0.2rem;
  color: #202B33;
  padding: 0.45rem 0.6rem;
  font-size: 0.9rem;
  font-family: inherit;
  height: 2rem;
  line-height: 1.125;
  -webkit-transition: background 0.2s,border 0.2s,box-shadow 0.2s,color 0.2s;
  transition: background 0.2s,border 0.2s,box-shadow 0.2s,color 0.2s;
}

.c2:focus {
  box-shadow: 0 0 0 0.1rem rgba(0,140,192,0.2);
  border-color: #008CC0;
}

.c2::-webkit-input-placeholder {
  color: #90A9BF;
}

.c2::-moz-placeholder {
  color: #90A9BF;
}

.c2:-ms-input-placeholder {
  color: #90A9BF;
}

.c2::placeholder {
  color: #90A9BF;
}

.c2:disabled,
.c2[disabled] {
  cursor: not-allowed;
  color: #90A9BF !important;
  background: #EFF3F6 !important;
}

.c2[type="file"] {
  height: auto;
}

.c3 {
  background: transparent;
  background-image: none;
  outline: none;
  max-width: 100%;
  display: block;
  position: relative;
  border: 0.05rem solid #BECEDC;
  border-radius: 0.2rem;
  color: #202B33;
  padding: 0.15rem 0.3rem;
  font-size: 0.7rem;
  font-family: inherit;
  height: 1.4rem;
  line-height: 1.125;
  -webkit-transition: background 0.2s,border 0.2s,box-shadow 0.2s,color 0.2s;
  transition: background 0.2s,border 0.2s,box-shadow 0.2s,color 0.2s;
}

.c3:focus {
  box-shadow: 0 0 0 0.1rem rgba(0,140,192,0.2);
  border-color: #008CC0;
}

.c3::-webkit-input-placeholder {
  color: #90A9BF;
}

.c3::-moz-placeholder {
  color: #90A9BF;
}

.c3:-ms-input-placeholder {
  color: #90A9BF;
}

.c3::placeholder {
  color: #90A9BF;
}

.c3:disabled,
.c3[disabled] {
  cursor: not-allowed;
  color: #90A9BF !important;
  background: #EFF3F6 !important;
}

.c3[type="file"] {
  height: auto;
}

.c0 h1 {
  font-size: 1.6rem;
}

.c0 h2 {
  font-size: 1.4rem;
}

.c0 h3 {
  font-size: 1.2rem;
}

.c0 h4 {
  font-size: 1rem;
}

.c0 h5 {
  font-size: 0.9rem;
}

.c0 h6 {
  font-size: 0.8rem;
}

.c0 h1,
.c0 h2,
.c0 h3,
.c0 h4,
.c0 h5,
.c0 h6 {
  font-weight: 700;
  line-height: 1.25;
  margin-bottom: 0.5em;
  margin-top: 0;
  font-family: Helvetica,Arial,sans-serif;
}

.c0 h1,
.c0 h2,
.c0 h3,
.c0 h4 {
  color: #394B59;
}

.c0 h5,
.c0 h6 {
  color: #293742;
}

.c0 p {
  color: #202B33;
  line-height: 1.5;
  margin-bottom: 0.5em;
  margin-top: 0;
  font-family: Helvetica,Arial,sans-serif;
}

.c0 a {
  font-family: Helvetica,Arial,sans-serif;
  color: #008CC0;
  outline: none;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 a:hover {
  color: #00A7E4;
}

.c0 a:focus,
.c0 a:hover,
.c0 a:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 a:focus,
.c0 a:active {
  color: #007099;
}

.c0 a:focus {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 b,
.c0 strong {
  font-weight: bolder;
}

.c0 small {
  font-size: 80%;
  font-weight: 400;
  color: #566E82;
}

.c0 blockquote {
  margin-left: 0;
  padding: 0.4rem 0.8rem;
  color: #566E82;
}

.c0 blockquote p:last-child {
  margin-bottom: 0;
}

.c0 table {
  border-collapse: collapse;
  border-spacing: 0;
  text-align: left;
  width: 100%;
}

.c0 th {
  border-bottom: 0.05rem solid #BECEDC;
  padding: 0.6rem 0.4rem;
  border-bottom-width: 0.1rem;
}

.c0 td {
  border-bottom: 0.05rem solid #EFF3F6;
  padding: 0.6rem 0.4rem;
}

<div>
  <div
    class="c0"
  >
    <input
      autocomplete="off"
      class="c1"
      placeholder="Default input"
      type="text"
      value=""
    />
    <input
      autocomplete="off"
      class="c2"
      placeholder="Large input"
      type="text"
      value=""
    />
    <input
      autocomplete="off"
      class="c3"
      placeholder="Small input"
      type="text"
      value=""
    />
  </div>
</div>
`;
